### PR TITLE
Add EventHandlerDropGuard

### PR DESCRIPTION
**Pull request by jplatte**
Creating such a guard basically makes the event handler temporary, it will be removed again once the guard is dropped – e.g. because of a surrounding future being cancelled, or an error being propagated out of a function that stored the guard on the stack.

cc @docweirdo

### CODE REVIEW

**Comment by jplatte** [comment]
Weird, GitHub complained about conflicts but rebasing was fully automatic.

File: crates/matrix-sdk/src/client/mod.rs [hunk 1]
```diff
@@ -677,20 +677,35 @@ impl Client {
...
     /// Remove the event handler associated with the handle.
     ///
-    /// Note that handlers that remove themselves will still execute
-    /// with events received in the same sync cycle.
+    /// Note that you **must not** call `remove_event_handler` from the
+    /// non-async part of an event handler, that is:
+    ///
+    /// ```ignore
+    /// client.add_event_handler(|ev: SomeEvent, client: Client, handle: EventHandlerHandle| {
+    ///     // ⚠ this will cause a deadlock ⚠
```
Comment (poljar) [inline crates/matrix-sdk/src/client/mod.rs:639]: Will this cause a deadlock? We do the following
1. Lock the event handlers map to read the handlers
2. Construct the futures for the handler callbacks
3. Release the lock
4. Run the futures in order.

We will try to remove the handler at step 4 which shouldn't hold on to the lock anymore, no?

Comment (jplatte) [reply crates/matrix-sdk/src/client/mod.rs:639]: No, this is step 2, below where there's another comment is step 4.

Comment (poljar) [reply crates/matrix-sdk/src/client/mod.rs:639]: Ah right, that's not part of the future, that's part of the method that constructs the future.

**Review by poljar — APPROVED** [review]
Looks good.

**Pull request merged**